### ISSUE TITLE
Optimizing JSON serialization/deserialization

### ISSUE THREAD

**Issue by lukehutch** [serverpod/serverpod#1675]
**Describe the bug**

Deserializing just two protocol objects on app startup adds ~1.2 seconds to my app startup time, according to profiling.

![image](https://github.com/serverpod/serverpod/assets/811305/e8751a5a-7043-4d8a-9d7b-2b540f3fdc36)

These objects consist of only maybe 150 fields total, maybe 2-3kb of JSON in total. It is really surprising that JSON deserialization is so slow.

I am going to parse these lazily, or asynchronously after the app starts up, but in general, this needs some major optimization work.

Serialization and deserialization is the bread and butter of Serverpod, so it needs to be very, very fast. Devilishly fast. This will affect scalability on the server too, since you're only running a single isolate right now, and as it is, when calling an endpoint that hits the database, you are running serialization and deserialization multiple times per request:

1. Deserializing the request from the app.
2. Serializing parameters for the database query.
3. Deserializing each query result row from whatever Postgres reutrns into a `Map`.
4. Re-serializing the `Map` into JSON to pass it back to the client.

I don't have a lot of time to dig into this, but I will add any observations I find here about what could be improved. Here is my first observation: `SerializationManage.deserialize` should be split out into a separate deserialization function for each deserializable type, to avoid the type comparison against multiple types in a single big if-statement. Type comparisons are expensive, and the caller already knows what type it expects. So the protocol class generation code can just generate code that calls the deserialization function for each specific type that is expected.

**Comment by Isakdl** [comment]
Thank you for the report! This is indeed quite an important piece in Serverpod as it is used a lot and needs to be fast. 

Just for my information, is your JSON files mostly flat or do you have many nested objects?

>  So the protocol class generation code can just generate code that calls the deserialization function for each specific type that is expected.

Discussed this point last week with the team and I agree that this probably is a better approach.

**Comment by lukehutch** [comment]
One protocol object is flat with a large number of columns (200ish?),while the other is a list of lists of strings (maybe 80 pairs of strings).

**Comment by lukehutch** [comment]
Just realized these numbers don't line up with my original estimates. Truth be told I haven't counted them. But I still consider these objects "small" as far as JSON documents go.

**Comment by Isakdl** [comment]
Okay so I setup a test on this, a test class with 150 fields with all different data types and with one field containing a self referencing list. 

Then created a massive recursive data structure with many deeply nested lists of this object. Then I measured the time to decode said object from a string, running our entire generated deserializer logic. The test was done with the benchmark package: [benchmark_harness](https://pub.dev/packages/benchmark_harness)

I get pretty consistent results for this at about 5.5k micro seconds or 0.0055 seconds. While I'm sure we can optimize the code I'm not getting anywhere near the results you have reported. 

I'm guessing there must be something else that is taking up time here, do you have any more insights into your setup? How many classes are you running with for example?

<img width="312" alt="Screenshot 2023-12-18 at 22 45 42" src="https://github.com/serverpod/serverpod/assets/7395515/df08693d-975a-49d4-ba0a-87e3b05d0a57">

**Comment by lukehutch** [comment]
Huh, well your numbers are more like what I would expect! I couldn't understand how this could be this slow.

There were only two protocol classes that were taking this much time to deserialize, according to profiling. I finally counted the columns: in the big (flat) protocol object, there are 235 columns, mostly `bool`, some `String` (with short strings in them, under 256 chars), and a few enums. The second object was 85 arrays of 4 strings each, in total less than 80 chars for each of the 85 arrays of arrays. It's really not a lot of data.

I already dramatically modified my code (made it all run in the background after app startup, and changed all fields to make them nullable, to reduce the amount that needs to be serialized). I should have made a git snapshot before I did this, but I don't think I can recreate this now.

Unfortunately the profiling tools for Dart really suck, so I wasn't able to get a good handle on exactly where the timing was being taken. And it's possible that the profiler just wasn't even recording the time correctly, for some reason (although you would think this would be a simple matter, given that Dart is single-threaded!).

I don't know what to say at this point... definitely serialization and deserialization should still be maximally optimized, but it looks like this problem is not as bad as I thought it was. Thanks for testing it!
